pkg/daemon/ceph/osd/remove.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osd
+
+import (
+	"fmt"
+	"github.com/prometheus/common/log"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+)
+
+// RemoveOSDs purges a list of OSDs from the cluster
+func RemoveOSDs(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdsToRemove []string) error {
+
+	// Generate the ceph config for running ceph commands similar to the operator
+	if err := writeCephConfig(context, clusterInfo); err != nil {
+		return errors.Wrap(err, "failed to write the ceph config")
+	}
+
+	osdDump, err := client.GetOSDDump(context, clusterInfo)
+	if err != nil {
+		return errors.Wrap(err, "failed to get osd dump")
+	}
+
+	for _, osdIDStr := range osdsToRemove {
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+		logger.Debugf("validating status of osd.%d", osdID)
+		status, _, err := osdDump.StatusByID(int64(osdID))
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to get osd status for osd %v", osdID))
+
+		}
+		const upStatus int64 = 1
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy. It cannot be removed unless it is 'down'", osdID)
+			continue
+		}
+		logger.Debugf("osd.%d is marked 'DOWN'. Removing it", osdID)
+		removeOSD(context, clusterInfo, osdID)
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) {
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+	// Mark the OSD as out.
+	output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
```

Comment:
Use `NewCephCommand()` instead of `context.Executor.ExecuteCommandWithOutput()`. The latter is only needed when not executing Ceph commands.